Add instancing support to DotScene Plugin (#2109)
OGRECave/ogre: Docs/src/manual.md

## AnimableValue

When implementing custom animable properties, you have to also implement a number of methods on the AnimableValue interface - basically anything which has been marked as unimplemented. These are not pure virtual methods simply because you only have to implement the methods required for the type of value you’re animating. Again, see the examples in Light to see how this is done.

@page Instancing Instancing

Instancing significantly reduces the CPU overhead of submitting many separate draw calls and is a great technique for rendering trees, rocks, grass, RTS units and other groups of similar (but necessarily identical) objects.

OGRE supports a variety of techniques to speed up the rendering of many objects in the Scene.

> <dl compact="compact">

Reviewer: this would be a more Doxygeny style:
```
@par @ref Static-Geometry
Pre-transforms and batches up meshes for efficient use as static geometry in a scene.

@par ..
``` 
with the same visual results

Author: I'm using the style of the other sections that use the same thing, I don't even know what it is 🤣

<dt>@ref Static-Geometry</dt> 
<dd>Pre-transforms and batches up meshes for efficient use as static geometry in a scene.</dd>
<dt>@ref Instance-Manager</dt> 
<dd>Instancing is a way of batching up geometry into a much more efficient form, but with some limitations, and still be able to move & animate it.</dd>
</dl>

@tableofcontents

# Static Geometry {#Static-Geometry}
Modern graphics cards (GPUs) prefer to receive geometry in large batches. 
It is orders of magnitude faster to render 10 batches of 10,000 triangles than it is to render 10,000 batches of 10 triangles, even though both result in the same number of on-screen triangles.

Therefore it is important when you are rendering a lot of geometry to batch things up into as few rendering calls as possible. 
This class allows you to build a batched object from a series of entities in order to benefit from this behaviour. Batching has implications of it's own though:

>  - Batched geometry cannot be subdivided; that means that the whole group will be displayed, or none of it will. This obivously has culling issues.

Reviewer: ```suggestion
 - A geometry region cannot be subdivided; that means that the whole group will be displayed, or none of it will. This obivously has culling issues.
```

 - A single world transform must apply to the entire batch. Therefore once you have batched things, you can't move them around relative to each other. 
   That's why this class is most useful when dealing with static geometry (hence the name). 
   In addition, geometry is effectively duplicated, so if you add 3 entities based on the same mesh in different positions, 
   they will use 3 times the geometry space than the movable version (which re-uses the same geometry). 
   So you trade memory and flexibility of movement for pure speed when using this class.
 - A single material must apply for each batch. In fact this class allows you to use multiple materials, 
   but you should be aware that internally this means that there is one batch per material. 
   Therefore you won't gain as much benefit from the batching if you use many different materials; try to keep the number down.

Consult the following links to get more information:

>  - API Documentation: [Ogre::StaticGeometry Class Reference](https://ogrecave.github.io/ogre/api/latest/class_ogre_1_1_static_geometry.html)

Reviewer: ```suggestion
 - API Documentation: [Ogre::StaticGeometry Class Reference](@ref Ogre::StaticGeometry)
```

Author: Thanks, I didn't know how to do that...

>  - Tutorial: [Tutorial - Static Geometry](https://ogrecave.github.io/ogre/api/latest/tut__static_geom.html)

Reviewer: ```suggestion
 - Tutorial: [Tutorial - Static Geometry](@ref tut_StaticGeom)
```

# Instance Manager {#Instance-Manager}
Instancing is a rendering technique to draw multiple instances of the same mesh using just one render call. There are two kinds of instancing:

Software: Two larges vertex & index buffers are created and the mesh vertices/indices are duplicated N number of times. When rendering, invisible instances receive a transform matrix filled with 0s. This technique can take a lot of VRAM and has limited culling capabilities.
Hardware: The hardware supports an extra param which allows Ogre to tell the GPU to repeat the drawing of vertices N number of times; thus taking considerably less VRAM. Because N can be controlled at runtime, individual instances can be culled before sending the data to the GPU.
Hardware techniques are almost always superior to Software techniques, but Software are more compatible, where as Hardware techniques require D3D9 or GL3, and is not supported in GLES2

All instancing techniques require shaders. It is not possible to use instancing with FFP (Fixed Function Pipeline)

Consult the following links to get more information:

>  - API Documentation: [Ogre::InstanceManager Class Reference](https://ogrecave.github.io/ogre/api/latest/class_ogre_1_1_instance_manager.html)

Reviewer: @ref Ogre::InstanceManager etc.

 - In-depth explanation: [What is instancing?](https://ogrecave.github.io/ogre/api/latest/instancing.html)

# Differences between both (#Differences-between-both)

> <table>

Reviewer: if you use markdown tables, this could be readable in the source-code as well. See:
https://github.com/OGRECave/ogre/blob/6c376059f03fbcbfd612e7cbf64216cf1b7ebd5b/OgreMain/include/OgreRenderSystem.h#L378

Author: Yes, but I thought that the doxygen tables look better.

Anyway, I don't even know how to generate the documentation with doxygen.

I installed it, the CMake detects it, but nothing is generated. It just copies the files. I'm trying to solve this before submitting again.

Reviewer: you have to run the "OgreDoc" target. e.g.
```
cmake --build . --target OgreDoc
```

Author: Thanks
